### PR TITLE
lets try to update node-semver

### LINKED ISSUE #6057: [feature] Use numerical version comparison in range resolver as long as version string is in the format a.b.c(.d.e)
We tried to use four-digit version numbers and resolve them with version range expressions. The handling of these (non-semver-compliant) version strings seems to be inconsistent in Conan. With Conan on Linux we observed that a version `3.8.0.1` is considered newer than a `3.8.0.0` when resolving the range `[*]`. Trying the same on Windows resolves always to `3.8.0.0`. There were also reports that for some users it would yield unexpected results on Linux as well, but I haven't confirmed that yet.

I think it would be beneficial if someone could look into this and rework the range resolver so that strictly numerical versions in the form a.b.c(.d.e...) still behave according to semver as far as the first three digits are concerned. Any extra digits (.d.e...) should be semantically equivalent to the patch version and compared numerically up to the first `-` or `+`. What do you think?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

### LINKED ISSUE #6459: change dependency on node-semver to be >= rather than ==
Hi, 
would it be possible to change the required dependency on node-semver to be >= 0.6.1 rather than ==0.6.1 or would this break things?

I have done quick but not exhaustive tests on a local build and can't see any problems.
Keeping this as is is something of an annoyance for distro packagers because they have to maintain a separate node-semver package just for conan. The same is true of patch-ng though I did get some errors when trying to unpin that so it may not be possible.

Thanks

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Fix: Upgrade node-semver python library.
Docs: omit

Fix https://github.com/conan-io/conan/issues/6057
Fix https://github.com/conan-io/conan/issues/6459

Related:
- https://github.com/conan-io/conan/issues/5366
- https://github.com/conan-io/conan/issues/2172

### CODE REVIEW

**Comment by memsharded** [comment]
An interesting question @jgsogo is if we should force 0.8.0 to ensure a consistent behavior of version-range resolution in different Conan installations

**Comment by jgsogo** [comment]
As it is a `0.x.y` version I think we should force a fixed version of the library. We don't know what can change in the next minor.

---

There other issues about this topic:
 * https://github.com/conan-io/conan/issues/5391
 * https://github.com/conan-io/conan/issues/5366

As you can read I vote to be deterministic even if the [SemVer documentation](https://semver.org/) doesn't consider microversions (or metadata). In fact, they have moved from SHOULD to MUST "ignore metadata" since them. Changing this means that it is no longer SemVer.

Is it breaking to move to something more deterministic? It wasn't because it depends on the order of the libraries retrieved from the server (was/is that order deterministic)?

**Comment by uilianries** [comment]
node-semver 0.8.0 changed the aspect when dealing to versions with more then 3-digits: https://github.com/podhmo/python-semver/issues/34

And what surprise, people asked because of Conan ...

The implementation is here: https://github.com/podhmo/python-semver/pull/35

What has changed? Now, comparing 3-digit to 4-digit will return as different versions. Before that, they were considered the same. Example: 1.1.1 == 1.1.1.1

**Comment by memsharded** [comment]
> What has changed? Now, comparing 3-digit to 4-digit will return as different versions. Before that, they were considered the same. Example: 1.1.1 == 1.1.1.1

Can we consider a bug that this happened in the past? Can you submit a fix to Conan to see what it means on our side? Thanks!

**Comment by memsharded** [comment]
Ok, we have a problem. We cannot upgrade to 0.8.0 because it does not support Python 2.7 anymore. Although Python 2.7 support in Conan is deprecated, it is not completely removed yet, and we want to keep it running at the moment. It seems this will need to wait until Conan 2.0?

**Comment by uilianries** [comment]
> Can we consider a bug that this happened in the past? Can you submit a fix to Conan to see what it means on our side? Thanks!

Yes, because we were allowing different versions when comparing them. e.g. 1.1.1 == 1.1.1.1 are different versions, but in the past it was true. If we use 0.8.0 we don't need a fix, because it's embedded in nose-semver. However, if we need to keep 0.6.0, we will need to fix it in Conan code.

> Ok, we have a problem. We cannot upgrade to 0.8.0 because it does not support Python 2.7 anymore. Although Python 2.7 support in Conan is deprecated, it is not completely removed yet, and we want to keep it running at the moment. It seems this will need to wait until Conan 2.0?

I see the perfect excuse for deprecating python 2.7 haha, well, let's keep 0.6.1 and fix it in Conan base code.

File: conans/requirements.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ PyYAML>=3.11, <6.0
 patch-ng>=1.17.4, <1.18
 fasteners>=0.14.1
 six>=1.10.0,<=1.14.0
-node-semver==0.6.1
+node-semver>=0.6.1, <0.8.0
```
Comment (uilianries) [inline conans/requirements.txt]:
```suggestion
node-semver>=0.6.1, <=0.8.0
```

@memsharded we shouldn't ignore 0.8.0. Both 0.6.1 and 0.8.0 will result in the same behavior.

Comment (memsharded) [reply conans/requirements.txt]: Yes, but we should probably put the limit on 0.8.0. As it is unstable 0.X, future versions can break. Please add <=0.8.0

File: conans/client/tools/version.py [hunk 2]
```diff
@@ -42,9 +46,14 @@ def build(self):
     def __eq__(self, other):
         if not isinstance(other, Version):
             other = Version(other)
-        return self._semver.compare(other._semver) == 0
+        return (self._semver.compare(other._semver) or self._compare_micro(other)) == 0
```
Comment (memsharded) [inline conans/client/tools/version.py:49]: So the key is that node-semver already had the ``_compare_micro()`` in 0.6.1?

Comment (uilianries) [reply conans/client/tools/version.py:49]: No, it has been included in 0.8.0. This is "port" for 0.6.1

File: conans/requirements.txt (hunk 1 above)
Comment (uilianries) [reply conans/requirements.txt]: Okay, @memsharded you can merge it directly by this comment

**Comment by jgsogo** [comment]
This implies that we no longer follow SemVer, with 4 numbers I've read it as "Explicit versioning" and it is not adding one number to the right, but one to left (non-breaking behavior is 1.2.X.Y), are we going to change package_id_modes?

If we are changing this I would suggest considering lexicographic order to build metadata as well, will it fit into _"Explicit versioning"_ or is it another just another different schema?

**Comment by memsharded** [comment]
> This implies that we no longer follow SemVer, with 4 numbers I've read it as "Explicit versioning" and it is not adding one number to the right, but one to left (non-breaking behavior is 1.2.X.Y), are we going to change package_id_modes?

Not sure what you mean with this, can you please clarify?

**Comment by uilianries** [comment]
I think it's because we have two Version classes, one in model, and one in tools. But this one is about tools, only for comparing project versions.

**Comment by jgsogo** [comment]
We are building a dependency manager, managing versions is one of our core features, we should think very carefully about any change here.

 * SemVer states that only MAJOR, MINOR, PATCH and pre-release have to be taken into account. Any other thing MUST not be taken into account. So this will no longer be SemVer.
 * Is there any SemVerXXX specification that includes MICRO versions, or are we on our own if we include it?
 * When I search on the Internet about 4-digit versions, I come across _"Explicit versioning"_ where the 4th number is not a MICRO version (added to the right: x.y.z.MICRO) but a DISRUPTIVE one (added to the left: DISRUPTIVE.x.y.z). Using this specification, packages are API-compatible if they share the first to numbers.

We are a package manager and we can follow one specification or we can create our own. But before breaking the one we are following right now, **we need to know where we are moving to**.

**Comment by uilianries** [comment]
@jgsogo Now I see your point. If we include a 4th digit we would be breaking the SemVer specification.

Checking in semver repo, I've found many people asking for 4th digit support, but this won't happen: https://github.com/semver/semver/issues/323#issuecomment-481450266

So considering it, node-semver is broken in 0.8.0

**Comment by SSE4** [comment]
4 digit is something very common, especially in Microsoft/Windows world. it's usually noted as **major.minor.build.private** (or **major.minor.build.revision**), and usually all resources, DLLs, etc. have these 4 version numbers. I think this is even part of PE/COFF specification.

**Comment by uilianries** [comment]
@SSE4 Kernel Linux also uses 4 digits, but it's not part of semver specification. The method tools.Version() uses semver under the hood.

The only close option is pre-release: https://semver.org/#spec-item-9

**Comment by memsharded** [comment]
> 4 digit is something very common, especially in Microsoft/Windows world. it's usually noted as major.minor.build.private (or major.minor.build.revision), and usually all resources, DLLs, etc. have these 4 version numbers. I think this is even part of PE/COFF specification.

The problem is that it might be a common thing for specifying versions. But I have been having a look too, and nothing out there defines any kind of version-range resolution against 4 digits versions. Some people might extrapolate the 4th version on the right, and make it a micro version, but this is not defined in the semver especification.

**Comment by SSE4** [comment]
indeed, this is not SemVer, and I think these are versioning schemes predating the SemVer. that particular 4-digit scheme appears to be very common, especially for old(er) projects, and such conventions are frequently used by companies in enterprise world. as a package manager serving both OSS and Enterprise cases, we should probably live with it, and support both versioning schemes, or maybe even arbitrary versioning schemes (e.g. if someone is using latin letters, or roman numbers, or cyrillic letters, or just arbitrary names). this may require us to: 1) expose several version classes (e.g. SemVer, FourDigitVer, etc.) 2) allow author to provide/expose his own comparators/operators to handle his versioning gracefully

**Comment by puetzk** [comment]
> I think this is even part of PE/COFF specification.

It's not part of PE/COFF, but it **is** part of the standard Win32 version resource, and it's also a required part of doing WinSxS manifest [assembly version](https://docs.microsoft.com/en-us/windows/win32/sbscs/assembly-versions) (where you *must* use 4 digits, if only by appending zeros). It's also CMake's default scheme for `PROJECT_VERSION` (where they call it major.minor.patch.tweak).

**Comment by memsharded** [comment]
This is very risky. Taking it for Conan 2.0

**Comment by memsharded** [comment]
This was too risky for Conan 1.X. Conan 2.0 in https://github.com/conan-io/conan/pull/10058 is removing the external dependency to node-semver, and implementing our own version and version-range management. Closing this.

**Pull request closed without merging**